File: p12.pl
% P12 (**) Decode a run-length encoded list.
% Given a run-length code list generated as specified in problem P11. Construct its uncompressed version.

decode([],[]).
% X is a single element.
decode([X|Xs],[X|Ys]) :-
    \+ is_list(X), % See also P07.
    decode(Xs,Ys).
% Only one left.
decode([[1,X]|Xs],[X|Ys]) :-
    decode(Xs,Ys).
% Add one X to Y until N = 1.
decode([[N,X]|Xs],[X|Ys]) :-
    N > 1,
    Nn is N - 1,
    decode([[Nn,X]|Xs],Ys).

File: quint/haskell/lists/p12.hs
-- P12 (**) Decode a run-length encoded list.

-- NOTE: See P11 if you want to use a data type.
decode :: Eq x => [Either x (Int, x)] -> [x]
decode x = decodeEither x []

decodeEither :: Eq x => [Either x (Int, x)] -> [x] -> [x]
decodeEither [] y               = y
decodeEither (Left x:xs) y      = decodeEither xs (y ++ [x])
decodeEither (Right (n,e):xs) y = decodeEither xs (y ++ (replicate n e))
